src/frontend/src/optimizer/rule/intersect_merge_rule.rs

```diff
@@ -0,0 +1,47 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::{BoxedRule, Rule};
+use crate::optimizer::plan_node::{LogicalIntersect, PlanTreeNode};
+use crate::optimizer::PlanRef;
+
+pub struct IntersectMergeRule {}
+impl Rule for IntersectMergeRule {
+    fn apply(&self, plan: PlanRef) -> Option<PlanRef> {
+        let top_intersect: &LogicalIntersect = plan.as_logical_intersect()?;
+        let top_all = top_intersect.all();
```

Comment:
I think it would be safer to check that it's not `intersect all`? This doesn't work for intersect all to me.

Reply:
In my opinion, this rule can be applicable to both "intersect" and "intersect all" operations, similar to "union" and "union all".